### ISSUE TITLE
Deadlock in AsyncCall when cancelling ongoing RPCs

### ISSUE THREAD

**Issue by Kalmalyzer** [grpc/grpc#14777]
I am experimenting with grpc-csharp in Unity, and I am encountering a deadlock during shutdown of my application. This happens with 30% probability when cancelling one particular RPC.

**Related bugs**
https://github.com/grpc/grpc/issues/9202 sounds related to this. I'm not sure -- but I think it is another bug.

**Background**
Some version information:
OS used: Windows 10 64-bit Home
Unity versions: 2017.3.0f3 and 2017.3.1f1
Unity Mono version: 4.6 (Experimental)
grpc-csharp version: 1.10.0 (fetched with NuGet)

I have created a backend service which manages connections for players. It has a set of RPCs which allow me to model a 'connection' from a game client to a backend, but over HTTP/2 instead of raw TCP (which is otherwise customary in realtime games).

Example service interface:
```
service PlayerConnectionService {

    rpc Connect (ConnectRequest) returns (ConnectResponse) {}
    rpc Ping (PingRequest) returns (PingResponse) {}
    rpc Disconnect (DisconnectRequest) returns (DisconnectResponse) {}
    
    rpc Disconnected (DisconnectedRequest) returns (stream DisconnectedNotification) {}
}
```

The backend is written in normal C#. The client is a Unity application.

**Problem scenario**
I call `Connect`, followed by `Disconnected`, followed by a number of other API methods. The Disconnected API will only provide notifications if/when the backend decides to end the session (for example, because the client hasn't sent any `Ping` RPCs lately).

All the other API methods seem to work well... but when I attempt to cancel the Disconnected RPC from the client side, I run into a deadlock in AsyncCall. It happens perhaps 30% of the time.

**Implementation details**
The Disconnected RPC is started by calling this method:
```
    public Task SubscribeDisconnectionAsync(string sessionId, DisconnectedNotificationCallbackDelegate notificationHandler, CancellationToken cancellationToken)
    {
        var call = GrpcClient.Disconnected(new PlayerConnectionService.DisconnectedRequest { SessionId = sessionId }, null, null, cancellationToken);

        Task listenerTask = Task.Run(async () =>
        {
            try
            {
                using (call)
                {
                    while (await call.ResponseStream.MoveNext(cancellationToken))
                    {
                        PlayerConnectionService.DisconnectedNotification disconnectedNotification = call.ResponseStream.Current;
                        notificationHandler(disconnectedNotification);
                    }
                }
            }
            catch (RpcException e) when (e.Status.StatusCode == StatusCode.Cancelled)
            {
                // Translate RpcException with StatusCode.Cancelled to a TaskCanceledException
                throw new TaskCanceledException();
            }
        });

        return listenerTask;
    }
```

The intention of `SubscribeDisconnectionAsync()` is that the calling site invokes it once, and when it is time to shut down (or disconnect manually), first send a cancellation and then `await` the listener task.

The cancellation is done like so:

```
    public void CancelAllTransactions()
    {
        <snip signalling of two other TokenCancellationSources>
        ...
        if (disconnectionCancellationTokenSource != null)
        {
            Debug.Log("Sending cancellation to disconnectionCancellationTokenSource");
            disconnectionCancellationTokenSource.Cancel();
            disconnectionCancellationTokenSource = null;
        }
    }
```

Now, there are thre different TaskCancellationSources being used. One of them is for a task that idles most of the time; another is for a bidirectional streaming RPC. For some reason, the deadlock only ever occurs within `disconnectionTaskCompletionSource.Cancel()`. I have no good explanation as to why.

When I break in with a debugger to inspect what's going on, I find two threads that are waiting on a mutex in AsyncCall:

```
>	Void Grpc.Core.Internal.AsyncCallBase`2:Cancel ()+0x0 at c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\AsyncCallBase.cs:[77:13-77:26]	C#	Symbols loaded.
 	Void Grpc.Core.Internal.AsyncCall`2:<RegisterCancellationCallback>b__34_0 ()+0x1 at c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\AsyncCall.cs:[417:70-417:83]	C#	Symbols loaded.
 	Void System.Threading.CancellationToken:ActionToActionObjShunt (Object)+0x6 at :-1	C#	No symbols loaded.
 	Void System.Threading.CancellationCallbackInfo:ExecutionContextCallback (Object)+0x13 at :-1	C#	No symbols loaded.
 	Void System.Threading.ExecutionContext:RunInternal (ExecutionContext, ContextCallback, Object, Boolean)+0x73 at :-1	C#	No symbols loaded.
 	Void System.Threading.ExecutionContext:Run (ExecutionContext, ContextCallback, Object, Boolean)+0x4 at :-1	C#	No symbols loaded.
 	Void System.Threading.ExecutionContext:Run (ExecutionContext, ContextCallback, Object)+0x2f at :-1	C#	No symbols loaded.
 	Void System.Threading.CancellationCallbackInfo:ExecuteCallback ()+0x2c at :-1	C#	No symbols loaded.
 	Void System.Threading.CancellationTokenSource:CancellationCallbackCoreWork (CancellationCallbackCoreWorkArguments)+0x43 at :-1	C#	No symbols loaded.
 	Void System.Threading.CancellationTokenSource:ExecuteCallbackHandlers (Boolean)+0xc3 at :-1	C#	No symbols loaded.
 	Void System.Threading.CancellationTokenSource:NotifyCancellation (Boolean)+0x61 at :-1	C#	No symbols loaded.
 	Void System.Threading.CancellationTokenSource:Cancel (Boolean)+0x8 at :-1	C#	No symbols loaded.
 	Void System.Threading.CancellationTokenSource:Cancel ()+0x2 at :-1	C#	No symbols loaded.
 	Void FreedomBackendClient:CancelAllTransactions ()+0x6f at C:\Plastic\FreedomBackend\UnityClient\Assets\FreedomBackendClient.cs:118	C#	Symbols loaded.
 	Void FreedomBackendClient:Disconnect (DisconnectDoneCallbackDelegate)+0x16 at C:\Plastic\FreedomBackend\UnityClient\Assets\FreedomBackendClient.cs:54	C#	Symbols loaded.
 	Void OnlineTestComponent:Disconnect ()+0x13 at C:\Plastic\FreedomBackend\UnityClient\Assets\OnlineTestComponent.cs:54	C#	Symbols loaded.
 	Void UnityEngine.Events.InvokableCall:Invoke ()+0x17 at C:\buildslave\unity\build\Runtime\Export\UnityEvent.cs:165	C#	Symbols loaded.
 	Void UnityEngine.Events.UnityEvent:Invoke ()+0x24 at C:\buildslave\unity\build\Runtime\Export\UnityEvent_0.cs:58	C#	Symbols loaded.
 	Void UnityEngine.UI.Button:Press ()+0x2d at C:\buildslave\unity\build\Extensions\guisystem\UnityEngine.UI\UI\Core\Button.cs:36	C#	Symbols loaded.
 	Void UnityEngine.UI.Button:OnPointerClick (PointerEventData)+0x12 at C:\buildslave\unity\build\Extensions\guisystem\UnityEngine.UI\UI\Core\Button.cs:45	C#	Symbols loaded.
 	Void UnityEngine.EventSystems.ExecuteEvents:Execute (IPointerClickHandler, BaseEventData)+0x8 at C:\buildslave\unity\build\Extensions\guisystem\UnityEngine.UI\EventSystem\ExecuteEvents.cs:50	C#	Symbols loaded.
 	Boolean UnityEngine.EventSystems.ExecuteEvents:Execute (GameObject, BaseEventData, EventFunction`1)+0x73 at C:\buildslave\unity\build\Extensions\guisystem\UnityEngine.UI\EventSystem\ExecuteEvents.cs:261	C#	Symbols loaded.
 	Void UnityEngine.EventSystems.StandaloneInputModule:ProcessMousePress (MouseButtonEventData)+0x17a at C:\buildslave\unity\build\Extensions\guisystem\UnityEngine.UI\EventSystem\InputModules\StandaloneInputModule.cs:554	C#	Symbols loaded.
 	Void UnityEngine.EventSystems.StandaloneInputModule:ProcessMouseEvent (Int32)+0x31 at C:\buildslave\unity\build\Extensions\guisystem\UnityEngine.UI\EventSystem\InputModules\StandaloneInputModule.cs:453	C#	Symbols loaded.
 	Void UnityEngine.EventSystems.StandaloneInputModule:ProcessMouseEvent ()+0x3 at C:\buildslave\unity\build\Extensions\guisystem\UnityEngine.UI\EventSystem\InputModules\StandaloneInputModule.cs:433	C#	Symbols loaded.
 	Void UnityEngine.EventSystems.StandaloneInputModule:Process ()+0x72 at C:\buildslave\unity\build\Extensions\guisystem\UnityEngine.UI\EventSystem\InputModules\StandaloneInputModule.cs:212	C#	Symbols loaded.
 	Void UnityEngine.EventSystems.EventSystem:Update ()+0xf9 at C:\buildslave\unity\build\Extensions\guisystem\UnityEngine.UI\EventSystem\EventSystem.cs:294	C#	Symbols loaded.
```
and
```
>	Void System.Threading.Thread:Sleep (Int32)+0x1a at :-1	C#	No symbols loaded.
 	Void System.Threading.SpinWait:SpinOnce ()+0x2d at :-1	C#	No symbols loaded.
 	Void System.Threading.CancellationTokenSource:WaitForCallbackToComplete (CancellationCallbackInfo)+0xc at :-1	C#	No symbols loaded.
 	Void System.Threading.CancellationTokenRegistration:Dispose ()+0x44 at :-1	C#	No symbols loaded.
 	Void Grpc.Core.Internal.AsyncCall`2:OnAfterReleaseResources ()+0x1f at c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\AsyncCall.cs:[331:13-331:54]	C#	Symbols loaded.
 	Void Grpc.Core.Internal.AsyncCallBase`2:ReleaseResources ()+0x1b at c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\AsyncCallBase.cs:[199:13-199:39]	C#	Symbols loaded.
 	Boolean Grpc.Core.Internal.AsyncCallBase`2:ReleaseResourcesIfPossible ()+0x47 at c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\AsyncCallBase.cs:[174:21-174:40]	C#	Symbols loaded.
 	Void Grpc.Core.Internal.AsyncCall`2:HandleFinished (Boolean, ClientSideStatus)+0x3d at c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\AsyncCall.cs:[507:17-507:46]	C#	Symbols loaded.
 	Void Grpc.Core.Internal.AsyncCall`2:Grpc.Core.Internal.IReceivedStatusOnClientCallback.OnReceivedStatusOnClient (Boolean, ClientSideStatus)+0x3 at c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\AsyncCall.cs:[536:13-536:53]	C#	Symbols loaded.
 	Void <>c:<.cctor>b__33_1 (Boolean, BatchContextSafeHandle, Object)+0xd at c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\CallSafeHandle.cs:[40:42-40:153]	C#	Symbols loaded.
 	Void Grpc.Core.Internal.BatchContextSafeHandle:Grpc.Core.Internal.IOpCompletionCallback.OnComplete (Boolean)+0x18 at c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\BatchContextSafeHandle.cs:[143:17-143:94]	C#	Symbols loaded.
 	Void Grpc.Core.Internal.GrpcThreadPool:RunCompletionQueueEventCallback (IOpCompletionCallback, Boolean)+0x2 at c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\GrpcThreadPool.cs:[232:17-232:46]	C#	Symbols loaded.
 	Void Grpc.Core.Internal.GrpcThreadPool:<.ctor>b__15_0 (Object)+0x8 at c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\GrpcThreadPool.cs:[71:90-71:161]	C#	Symbols loaded.
 	Void System.Threading.QueueUserWorkItemCallback:WaitCallback_Context (Object)+0x13 at :-1	C#	No symbols loaded.
 	Void System.Threading.ExecutionContext:RunInternal (ExecutionContext, ContextCallback, Object, Boolean)+0x73 at :-1	C#	No symbols loaded.
 	Void System.Threading.ExecutionContext:Run (ExecutionContext, ContextCallback, Object, Boolean)+0x4 at :-1	C#	No symbols loaded.
 	Void System.Threading.QueueUserWorkItemCallback:System.Threading.IThreadPoolWorkItem.ExecuteWorkItem ()+0x2e at :-1	C#	No symbols loaded.
 	Boolean System.Threading.ThreadPoolWorkQueue:Dispatch ()+0x75 at :-1	C#	No symbols loaded.
 	Boolean System.Threading._ThreadPoolWaitCallback:PerformWaitCallback ()+0x0 at :-1	C#	No symbols loaded.
```

**Analysis**

I have dug through the gRPC codebase a bit, and this looks like a deadlock has occurred, centered around `AsyncCallBase.myLock`:

- The main thread has invoked CancellationTokenSource.Cancel().
- The main thread is the first to call cancel, and will therefore invoke CancellationTokenSource.ExecuteCallbackHandlers().
- For a particular registered handler, there is no TargetSyncContext, so CancellationCallbackCoreWork() is called on the current thread.
- CancellationCallbackInfo.ExecuteCallback() notes that there is a TargetExecutionContext.
- It activates the TargetExecutionContext and invokes CancellationCallbackInfo.ExecutionContextCallback().
- This, in turn, invokes CancellationToken.ActionToActionObjShunt() which is a thin wrapper that calls the following Lambda in AsyncCall.cs:417:
	() => this.Cancel()
- The lambda leads to AsyncCallBase.Cancel() being invoked. This grabs `myLock` - at which point the main thread will freeze, because the other thread already has acquired `myLock` in AsyncCall.HandleFinished(). The other thread is meanwhile waiting for the main thread to complete the Cancel() call. Deadlock ensues.

I think part of the problem is that INativeCall.Dispose() is calling out to other code while holding a mutex.

No good ideas what to do about it. It took me a day just to be 90% sure this wasn't just a problem when using Grpc in Unity, but a .NET-implementation-agnostic logic error / race condition.

**Comment by Kalmalyzer** [comment]
I will try to make a minimal repro case of this (Unity client, C# backend), ETA next week.

**Comment by Kalmalyzer** [comment]
Pardon me - I never got around to creating a proper repro case, and we are moving away from using gRPC in our current project. I won't be able to produce any repro case here. Feel free to close / tag as "no repro" as you wish.

**Comment by jtattermusch** [comment]
@Kalmalyzer  sorry to hear that, please consider filling out the gRPC survey https://grpc.io/blog/take-the-grpc-survey to provide us some feedback. And thanks for filing the issue anyway!
Also, please keep in mind gRPC C# doesn't officially support Unity yet - the support is experimental (see https://github.com/grpc/grpc/blob/master/src/csharp/experimental/README.md) for now so seeing issues when using it is not all that surprising.

**Comment by jtattermusch** [comment]
The analysis seems to be correct.
The problem seems to be we are calling `cancellationTokenRegistration?.Dispose();`  which can actually block in `System.Threading.CancellationTokenSource:WaitForCallbackToComplete` from under a lock (in e.g. in HandleFinished), but cancellationTokenSource.Cancel() also grabs the lock by invoking call.Cancel().

https://github.com/grpc/grpc/blob/5d302d1a9fd3afb486614ec34cdf643f4c5d53ac/src/csharp/Grpc.Core/Internal/AsyncCall.cs#L331

**Comment by jtattermusch** [comment]
Should be fixed by #16440 (which will be in the upcoming 1.15.0 release)
